File: focus_todo_mobile/lib/sync/sync_coordinator.dart
import '../shared/models/task.dart';
import '../shared/models/task_sync_metadata.dart';
import 'canonical_task_mapper.dart';
import 'merge_engine.dart';
import 'outbound_selector.dart';
import 'sync_metadata_helper.dart';

class SyncCoordinatorState {
  const SyncCoordinatorState({
    required this.initializedAfterHydration,
    required this.tasks,
    required this.syncMetadata,
  });

  final bool initializedAfterHydration;
  final List<Task> tasks;
  final List<TaskSyncMetadata> syncMetadata;
}

abstract class SyncCoordinator {
  SyncCoordinatorState initializeAfterHydration({
    required List<Task> tasks,
    required List<TaskSyncMetadata> syncMetadata,
  });

  List<CanonicalTask> prepareOutboundBatch(SyncCoordinatorState state);

  SyncCoordinatorState applyInboundBatch(
    SyncCoordinatorState state,
    List<CanonicalTask> mockData,
  );
}

class TaskSyncCoordinator extends MockSyncCoordinator {
  const TaskSyncCoordinator();
}

const SyncCoordinator syncCoordinator = TaskSyncCoordinator();

bool _sameWinner(Task left, Task right) {
  return left.schemaVersion == right.schemaVersion &&
      left.id == right.id &&
      left.title == right.title &&
      left.note == right.note &&
      left.dueDate == right.dueDate &&
      left.priority == right.priority &&
      left.category == right.category &&
      left.completed == right.completed &&
      left.completedAt == right.completedAt &&
      left.createdAt == right.createdAt &&
      left.updatedAt == right.updatedAt &&
      left.deletedAt == right.deletedAt &&
      left.createdByDeviceId == right.createdByDeviceId &&
      left.updatedByDeviceId == right.updatedByDeviceId;
}

class MockSyncCoordinator implements SyncCoordinator {
  const MockSyncCoordinator();

  @override
  SyncCoordinatorState initializeAfterHydration({
    required List<Task> tasks,
    required List<TaskSyncMetadata> syncMetadata,
  }) {
    return SyncCoordinatorState(
      initializedAfterHydration: true,
      tasks: List<Task>.from(tasks, growable: false),
      syncMetadata: List<TaskSyncMetadata>.from(syncMetadata, growable: false),
    );
  }

  @override
  List<CanonicalTask> prepareOutboundBatch(SyncCoordinatorState state) {
    if (!state.initializedAfterHydration) {
      return const <CanonicalTask>[];
    }

    return getOutboundTasks(
      state.tasks,
      state.syncMetadata,
    ).map(toCanonicalTask).toList(growable: false);
  }

  @override
  SyncCoordinatorState applyInboundBatch(
    SyncCoordinatorState state,
    List<CanonicalTask> mockData,
  ) {
    if (!state.initializedAfterHydration || mockData.isEmpty) {
      return SyncCoordinatorState(
        initializedAfterHydration: state.initializedAfterHydration,
        tasks: List<Task>.from(state.tasks, growable: false),
        syncMetadata: List<TaskSyncMetadata>.from(
          state.syncMetadata,
          growable: false,
        ),
      );
    }

    var tasks = List<Task>.from(state.tasks, growable: true);
    var syncMetadata = List<TaskSyncMetadata>.from(
      state.syncMetadata,
      growable: false,
    );

    for (final CanonicalTask item in mockData) {
      final remoteTask = fromCanonicalTask(item);
      final localIndex = tasks.indexWhere(
        (Task task) => task.id == remoteTask.id,
      );

      if (localIndex == -1) {
        tasks.add(remoteTask);
        syncMetadata = markSynced(
          syncMetadata,
          remoteTask.id,
          options: MarkSyncedOptions(
            lastKnownServerUpdatedAt: remoteTask.updatedAt,
          ),
        );
        continue;
      }

      final localTask = tasks[localIndex];
      final mergedTask = mergeTask(localTask, remoteTask);
      tasks[localIndex] = mergedTask;

      syncMetadata = _sameWinner(mergedTask, remoteTask)
          ? markSynced(
              syncMetadata,
              remoteTask.id,
              options: MarkSyncedOptions(
                lastKnownServerUpdatedAt: remoteTask.updatedAt,
              ),
            )
          : mergedTask.deletedAt != null
          ? markPendingDelete(syncMetadata, mergedTask.id)
          : markPendingUpsert(syncMetadata, mergedTask.id);
    }

    return SyncCoordinatorState(
      initializedAfterHydration: state.initializedAfterHydration,
      tasks: List<Task>.unmodifiable(tasks),
      syncMetadata: List<TaskSyncMetadata>.unmodifiable(syncMetadata),
    );
  }
}

File: focus_todo_computer/focus-todo/src/sync/syncCoordinator.ts
import type { Task, TaskSyncMetadata } from '../types/task'
import {
  fromCanonicalTask,
  toCanonicalTask,
  type CanonicalTask,
} from './canonicalTaskMapper'
import { mergeTask } from './mergeEngine'
import { getOutboundTasks } from './outboundSelector'
import {
  markError,
  markPendingDelete,
  markPendingUpsert,
  markSynced,
} from './syncMetadata'
import type { SyncWriteResult } from './syncTransport'

export interface SyncCoordinatorState {
  initializedAfterHydration: boolean
  tasks: Task[]
  syncMetadata: TaskSyncMetadata[]
}

export interface SyncCoordinator {
  initializeAfterHydration(input: {
    tasks: Task[]
    syncMetadata: TaskSyncMetadata[]
  }): SyncCoordinatorState
  prepareOutboundBatch(state: SyncCoordinatorState): CanonicalTask[]
  applyWriteResult(
    state: SyncCoordinatorState,
    taskId: string,
    result: SyncWriteResult,
  ): SyncCoordinatorState
  applyInboundBatch(
    state: SyncCoordinatorState,
    mockData: CanonicalTask[],
  ): SyncCoordinatorState
}

function sameWinner(left: Task, right: Task): boolean {
  return (
    left.schemaVersion === right.schemaVersion &&
    left.id === right.id &&
    left.title === right.title &&
    left.note === right.note &&
    left.dueDate === right.dueDate &&
    left.priority === right.priority &&
    left.category === right.category &&
    left.completed === right.completed &&
    left.completedAt === right.completedAt &&
    left.createdAt === right.createdAt &&
    left.updatedAt === right.updatedAt &&
    left.deletedAt === right.deletedAt &&
    left.createdByDeviceId === right.createdByDeviceId &&
    left.updatedByDeviceId === right.updatedByDeviceId
  )
}

export const syncCoordinator: SyncCoordinator = {
  initializeAfterHydration(input) {
    return {
      initializedAfterHydration: true,
      tasks: [...input.tasks],
      syncMetadata: [...input.syncMetadata],
    }
  },

  prepareOutboundBatch(state) {
    if (!state.initializedAfterHydration) return []
    return getOutboundTasks(state.tasks, state.syncMetadata).map(toCanonicalTask)
  },

  applyWriteResult(state, taskId, result) {
    if (!result.remoteTask) {
      return {
        initializedAfterHydration: state.initializedAfterHydration,
        tasks: [...state.tasks],
        syncMetadata: markError(
          state.syncMetadata,
          taskId,
          'Missing canonical row in sync write result',
        ),
      }
    }

    const remoteTask = fromCanonicalTask(result.remoteTask)
    const nextUpdatedAt = result.rowUpdatedAt ?? result.remoteTask.updated_at
    const existingLocalTask = state.tasks.find((task) => task.id === taskId)

    if (!existingLocalTask || result.applied) {
      return {
        initializedAfterHydration: state.initializedAfterHydration,
        tasks: [
          remoteTask,
          ...state.tasks.filter((task) => task.id !== remoteTask.id),
        ],
        syncMetadata: markSynced(state.syncMetadata, remoteTask.id, {
          lastKnownServerUpdatedAt: nextUpdatedAt,
        }),
      }
    }

    const mergedTask = mergeTask(existingLocalTask, remoteTask)

    return {
      initializedAfterHydration: state.initializedAfterHydration,
      tasks: state.tasks.map((task) =>
        task.id === mergedTask.id ? mergedTask : task,
      ),
      syncMetadata: sameWinner(mergedTask, remoteTask)
        ? markSynced(state.syncMetadata, remoteTask.id, {
            lastKnownServerUpdatedAt: nextUpdatedAt,
          })
        : mergedTask.deletedAt
          ? markPendingDelete(state.syncMetadata, mergedTask.id)
          : markPendingUpsert(state.syncMetadata, mergedTask.id),
    }
  },

  applyInboundBatch(state, mockData) {
    if (!state.initializedAfterHydration || mockData.length === 0) {
      return {
        initializedAfterHydration: state.initializedAfterHydration,
        tasks: [...state.tasks],
        syncMetadata: [...state.syncMetadata],
      }
    }

    let tasks = [...state.tasks]
    let syncMetadata = [...state.syncMetadata]

    for (const item of mockData) {
      const remoteTask = fromCanonicalTask(item)
      const localTask = tasks.find((task) => task.id === remoteTask.id)
      if (!localTask) {
        tasks = [...tasks, remoteTask]
        syncMetadata = markSynced(syncMetadata, remoteTask.id, {
          lastKnownServerUpdatedAt: item.updated_at,
        })
        continue
      }

      const mergedTask = mergeTask(localTask, remoteTask)
      tasks = tasks.map((task) => (task.id === mergedTask.id ? mergedTask : task))

      syncMetadata = sameWinner(mergedTask, remoteTask)
        ? markSynced(syncMetadata, remoteTask.id, {
            lastKnownServerUpdatedAt: item.updated_at,
          })
        : mergedTask.deletedAt
          ? markPendingDelete(syncMetadata, mergedTask.id)
          : markPendingUpsert(syncMetadata, mergedTask.id)
    }

    return {
      initializedAfterHydration: state.initializedAfterHydration,
      tasks,
      syncMetadata,
    }
  },
}
